File: peer/src/consumer.rs
use log::*;
use std::{
    cmp,
    collections::HashMap,
    error::Error,
    sync::{Arc, Mutex},
};

use crate::PeerContext;
use common::{
    types::*,
    util::{print_cdoc, system_time_msec}, default::PEER_LOG_SIZE,
};
use dittolive_ditto::prelude::*;

pub struct PeerConsumer {
    local_id: PeerId,
    last_ts_idx_by_peer: HashMap<PeerId, (u64, u32)>,
    msg_latency: LatencyStats,
    msg_latency_total: u64,
    // To keep subscription alive as needed
    #[allow(dead_code)]
    subscription: Subscription,
    pub live_query: Option<LiveQuery>,
}

fn incr_wrap(i: u32, max: u32) -> u32 {
    let mut r = i + 1;
    if r > max {
        r = 0;
    }
    r
}

impl PeerConsumer {
    fn new(local_id: PeerId, subscription: Subscription) -> Self {
        Self {
            local_id,
            last_ts_idx_by_peer: HashMap::new(),
            msg_latency: LatencyStats::new(),
            msg_latency_total: 0,
            subscription,
            live_query: None,
        }
    }

    // get timestamp of last record consumed, and expected next index
    fn get_ts_idx(&self, peer_id: &PeerId) -> (u64, u32) {
        let r = self.last_ts_idx_by_peer.get(peer_id);
        r.unwrap_or(&(0,0)).clone()
    }

    // set last timestamp and log index we consumed for this peer
    fn set_consumed_ts_idx(&mut self, peer_id: PeerId, ts: u64, mut i: u32, max_i: u32) {
        i = incr_wrap(i, max_i);
        self.last_ts_idx_by_peer.insert(peer_id, (ts, i));
    }

    fn process_peer(&mut self, id: PeerId, pl: &PeerLog) {
        let now = system_time_msec();
        let (mut ts, mut i) = self.get_ts_idx(&id);
        debug!("--> process_peer {} w/ log len {}", id, pl.log.len());
        loop {
            let rec = pl.log.get(i.to_string().as_str());
            if rec.is_none() || rec.unwrap().timestamp < ts {
                // we either got no record, or have wrapped to an old one
                break;
            }
            let r = rec.unwrap();
            let latency = now - r.timestamp;
            self.msg_latency_total += latency;
            self.msg_latency.num_events += 1;
            self.msg_latency.min_msec = cmp::min(self.msg_latency.min_msec, latency);
            self.msg_latency.max_msec = cmp::max(self.msg_latency.max_msec, latency);
            self.msg_latency.avg_msec = self.msg_latency_total / self.msg_latency.num_events;
            debug!("--> got peer record {:?} w/ latency {}", r, latency);
            i = incr_wrap(i, PEER_LOG_SIZE-1);
            ts = r.timestamp
        }
        self.set_consumed_ts_idx(id, ts, i, PEER_LOG_SIZE-1);
    }

    fn process_peer_doc(&mut self, pdoc: &PeerDoc) {
        for (peer_id, log) in &pdoc.logs {
            if peer_id == &self.local_id {
                // don't process your own records
                continue;
            }
            self.process_peer(peer_id.to_string(), log);
        }
    }

    pub fn get_message_latency(&self) -> LatencyStats {
        let mut stats = self.msg_latency.clone();
        stats.distinct_peers = self.last_ts_idx_by_peer.len();
        stats
    }
}

pub fn consumer_create_collection(pctx: &PeerContext) -> Result<Collection, Box<dyn Error>> {
    let store = pctx.ditto.store();
    let plan = pctx.get_plan().unwrap();
    let cc = store.collection(&plan.peer_collection_name)?;
    let mut logs = HashMap::new();
    logs.insert(pctx.id.clone(), PeerLog::new());

    let doc = PeerDoc {
        _id: plan.peer_doc_id.clone(),
        logs,
    };
    cc.upsert(doc)?;
    Ok(cc)
}

pub type PeerConsumerRef = Arc<Mutex<PeerConsumer>>;

pub fn consumer_start(pctx: &PeerContext) -> Result<PeerConsumerRef, Box<dyn Error>> {
    let coll = pctx.peer_collection.as_ref().unwrap().lock();
    let plan = pctx.get_plan().unwrap();
    let peer_doc_id = plan.peer_doc_id.clone();
    let query = coll.as_ref().unwrap().find_by_id(&peer_doc_id);
    info!(
        "--> consumer_start for coll {} w/ doc id {}",
        plan.peer_collection_name,
        peer_doc_id.to_query_compatible(StringPrimitiveFormat::WithoutQuotes)
    );

    let _consumer = Arc::new(Mutex::new(PeerConsumer::new(pctx.id.clone(), query.subscribe())));
    let consumer = _consumer.clone();
    let live_query = query
        .observe_local(move |doc: Option<BoxedDocument>, event| {
            trace!("-> observe peer event {:?}", event);
            if doc.is_none() {
                return;
            }
            let r = doc.as_ref().unwrap().typed::<PeerDoc>();
            match r {
                Ok(pdoc) => {
                    //let p = doc.unwrap().to_cbor().unwrap();
                    //print_cdoc(&p).unwrap();
                    consumer.lock().unwrap().process_peer_doc(&pdoc);
                }
                Err(e) => {
                    error!("PeerDoc deser Error {:?}", e);
                    let p = doc.unwrap().to_cbor().unwrap();
                    info!("received peer doc:");
                    print_cdoc(&p).unwrap();
                }
            }
        })
        .unwrap();
    _consumer.lock().unwrap().live_query = Some(live_query);
    Ok(_consumer)
}

File: ts/src/consumer.ts
import { Collection, DocumentPath, LiveQuery, Subscription, DocumentID } from '@dittolive/ditto'
import { PeerContext } from './context.js'
import { LatencyStats, PeerId, PeerDoc, PeerLogs, PeerLog, PeerRecord } from './types.js'
import { system_time_msec } from './util.js'
import { PEER_LOG_SIZE } from './default.js'

interface TimestampIndex {
    ts: number
    i: number
}

export class Consumer {
    pctx: PeerContext
    running: boolean
    lastTsIdxByPeer: Map<PeerId, TimestampIndex>
    msgLatency: LatencyStats
    msgLatencyTotal: number
    collection: Collection
    liveQuery: LiveQuery | null
    timeout: NodeJS.Timeout | null
    sub: Subscription | null


    constructor(pctx: PeerContext, peerColl: Collection) {
        this.pctx = pctx
        this.running = false
        this.lastTsIdxByPeer = new Map<PeerId, TimestampIndex>()
        this.msgLatency = new LatencyStats()
        this.msgLatencyTotal = 0
        this.collection = peerColl
        this.liveQuery = null
        this.timeout = null
        this.sub = null
    }

    async createPeerCollection(): Promise<Collection> {
        const store = this.pctx.ditto!.store
        const plan = this.pctx.coord_info!.execution_plan!
        const pc = store.collection(plan.peer_collection_name)
        const log : {[key: string]: PeerRecord} = {}
        const logs : {[key: string]: PeerLog} = {}
        logs[this.pctx.id] = { log: log }
        const pdoc: PeerDoc = { _id: plan.peer_doc_id, logs: logs }
        await pc.upsert(pdoc)
        return pc
    }

    async processPeer(pid: PeerId, plog: PeerLog): Promise<void> {
        const now = system_time_msec()
        let {ts, i} = this.getTSIdx(pid)
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const rec = plog.log[i.toString()]
            if (!rec || rec.timestamp < ts) {
                // no record, or we wrapped around to an old one
                break;
            }
            const latency = now - rec.timestamp
            this.msgLatencyTotal += latency
            this.msgLatency.num_events += 1
            this.msgLatency.min_msec = Math.min(this.msgLatency.min_msec, latency)
            this.msgLatency.max_msec = Math.max(this.msgLatency.max_msec, latency)
            this.msgLatency.avg_msec = this.msgLatencyTotal / this.msgLatency.num_events
            console.debug('--> got peer record ', rec, ` w/ latency ${latency}`)
            i = incrWrap(i, PEER_LOG_SIZE-1)
            ts = rec.timestamp
        }
        this.setConsumedTSIdx(pid, ts, i, PEER_LOG_SIZE-1)
    }

    // get timestamp of last record consumed, and expected next index
    getTSIdx(pid: PeerId): TimestampIndex {
        let r = this.lastTsIdxByPeer.get(pid)
        if (!r) {
            r = { ts: 0, i: 0 }
        }
        return r
    }

    // set last timestamp and log index we consumed for this peer
    setConsumedTSIdx(pid: PeerId, ts: number, idx: number, maxIdx: number) {
        const i = incrWrap(idx, maxIdx)
        this.lastTsIdxByPeer.set(pid, { ts: ts, i: i })
    }

    // Start consumer and return the peer collection
    async start(): Promise<Collection> {
        const plan = this.pctx.coord_info!.execution_plan!
        const docId: DocumentID = plan.peer_doc_id
        const docIdStr = docId.toQueryCompatibleString
        console.info(`--> consumer start for coll ${plan.peer_collection_name}` +
                     `w/ doc id ${docIdStr} logged: `, docId)
        this.running = true
        const pc = await this.createPeerCollection()
        const query = pc.findByID(plan.peer_doc_id)
        this.sub = query.subscribe()
        /* XXX TODO use DQL
        this.timeout = setInterval(async () => {
            const result = await this.pctx.ditto!.store.execute(
                `SELECT * FROM COLLECTION "${plan.peer_collection_name}" (logs MAP) WHERE _id = "${plan.peer_doc_id}"`)
            for (const item of result.items) {
                console.debug(`--> query result ${item.jsonString()}`)
                console.debug(`--> query result ${stringify(item)}`)
            }
        }, 1000)
        */

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        this.liveQuery = query.observeLocal(async (doc, _event) => {
            if (!this.running) {
                return
            }
            if (doc) {
                const logsPath: DocumentPath = doc.at('logs')
                const logs: PeerLogs = logsPath.value
                if (!logs) {
                    return;
                }
                for (const peerId of Object.keys(logs)) {
                    if (peerId == this.pctx.id) {
                        continue
                    }
                    this.processPeer(peerId, logs[peerId])
                }
            }
        })
        return pc
    }

    async stop(): Promise<LatencyStats> {
        console.info("--> consumer stop")
        this.running = false
        if (this.liveQuery) {
            this.liveQuery!.stop()
        }
        if (this.timeout) {
            clearInterval(this.timeout)
        }
        this.msgLatency.distinct_peers = this.lastTsIdxByPeer.size
        return this.msgLatency
    }

}

function incrWrap(i: number, max: number): number {
    let r = i + 1
    if (r > max) {
        r = 0
    }
    return r
}
